File: AdventOfCode2021/d13/d13.hs
import Control.Monad (join)
import Data.Bifunctor (bimap)
import Data.Set (Set, empty, fromList, size, member)
import qualified Data.Set as Set (map)

splitLs :: (Eq a) => [a] -> a -> ([a], [a])
splitLs ls elt = (\(a,b) -> (a, tail b)) $ span (/= elt) ls

readDots :: String -> (Int, Int)
readDots str = join bimap read $ splitLs str ','

readFolds :: String -> (String, Int)
readFolds s = bimap (drop (length "fold along ")) read $ splitLs s '='

type Input = ([(Int, Int)], [(String, Int)])
procInp :: String -> Input
procInp x = bimap (map readDots) (map readFolds) $ splitLs (lines x) ""

doFold brd v idx = (flip Set.map) brd
                   (\(x,y) -> if idx == 0
                              then if x < v then (x,y) else (2 * v - x, y)
                              else if y < v then (x,y) else (x, 2 * v - y))

fold brd axis v = if axis == "x" then doFold brd v 0 else doFold brd v 1

minMaxRange ls = [(minimum ls)..(maximum ls)]

printGrid brd = unlines [[if member (x,y) brd then '#' else ' ' | x <- xs ] | y <- ys]
    where xs = minMaxRange $ Set.map (\(x,_) -> x) brd
          ys = minMaxRange $ Set.map (\(_,y) -> y) brd

partA (dots, folds) = size $ uncurry (fold (fromList dots)) $ head folds

partB (dots, folds) = printGrid $ foldl (\brd f -> uncurry (fold brd) $ f) (fromList dots) folds

runProg s = unlines ["Part 1: " ++ (show $ partA inp), "Part 2:"] ++ (partB inp)
    where inp = procInp s

main = interact runProg

File: AdventOfCode2021/pysolns/d13.py
import sys

def lines():
  return [line.strip() for line in sys.stdin]

ls = lines()
dots = []
while ls and ls[0] != "":
  dots.append([int(x) for x in ls[0].split(",")])
  ls = ls[1:]
obrd = {(x,y) for x,y in dots}

folds = [(lambda x: (x[0], int(x[1])))(l.split()[2].split('=')) for l in ls[1:]]

def dofold(brd, v, idx):
  b = set()
  for pt in brd:
    if pt[idx] < v:
      b.add(pt)
      continue
    b.add((2*v - pt[0], pt[1]) if idx == 0 else (pt[0], 2*v - pt[1]))
  return b

def fold(brd, f):
    return dofold(brd, f[1], 0 if f[0] == 'x' else 1)

def partA():
  print(len(fold(obrd, folds[0])))

def partB():
  b = obrd
  for f in folds:
    b = fold(b, f)

  mnx = min(x for x,_ in b)
  mxx = max(x for x,_ in b)
  mny = min(y for _,y in b)
  mxy = max(y for _,y in b)
  for y in range(mny, mxy + 1):
    print(''.join('#' if (x,y) in b else ' ' for x in range(mnx, mxx + 1)))

partA()
partB()
